### PR TITLE
[PR]using read_exact instread of read when read length buffer.

### PR DESCRIPTION
**Pull request by Leoyzen**
#3 
rust-lang/rust#22570

### CODE REVIEW

File: src/io.rs [hunk 1]
```diff
@@ -39,12 +40,11 @@ pub mod async_ {
...
         let len_buf = {
             let mut len_buf = [0u8; std::mem::size_of::<u64>()];
-            match reader.read(&mut len_buf).await {
-                Ok(0) => return Ok(None),
-                Ok(n) if n == len_buf.len() => (),
-                Ok(_) => return Err(Error::UnexpectedEofError),
-                Err(error) => return Err(error.into()),
-            };
+            match reader.read_exact(&mut len_buf).await {
+                Err(ref e) if e.kind() == ErrorKind::UnexpectedEof => return Ok(None),
```
Comment (jerry73204) [inline src/io.rs:44]: It does not distinguish whether length is zero or not. They have different meanings here. If the length is zero, it hits an EOF with no more data, but if length is not zero, it ends with incomplete data. The function has to tell caller if there is more data available and whether it is complete.

File: src/io.rs [hunk 2]
```diff
@@ -149,11 +149,10 @@ pub mod blocking {
     {
         let len_buf = {
             let mut len_buf = [0u8; std::mem::size_of::<u64>()];
-            match reader.read(&mut len_buf) {
-                Ok(0) => return Ok(None),
-                Ok(n) if n == len_buf.len() => (),
-                Ok(_) => return Err(Error::UnexpectedEofError),
-                Err(error) => return Err(error.into()),
+            match reader.read_exact(&mut len_buf) {
+                Err(ref e) if e.kind() == ErrorKind::UnexpectedEof => return Ok(None),
```
Comment (jerry73204) [inline src/io.rs:153]: Same as above.

**Comment by jerry73204** [comment]
Yes, it deserves a fix. `read()` is not guaranteed to reach expected length, even there are remaining in file or on device. The nuance part is that reading zero length has a special meaning. It indicates the end of file.

The `try_read_len()` returns `Ok(None)` if it hits immediate EOF. If it reads remaining bytes, it returns compete data or raises error when hitting early EOF. I put some review to reflect this change.

**Review by jerry73204 — CHANGES_REQUESTED** [review]
Please distinguish the cases if length is zero or not. They have different meaning here.

File: src/io.rs (hunk 1 above)
Comment (Leoyzen) [reply src/io.rs:44]: @jerry73204 It's quitely hard to distinguish these two situation. see this: [read_exact and EOF](https://internals.rust-lang.org/t/read-exact-and-eof/3047/2).
I think it's normal behavior just quit the processing when there is breaking. I also saw this from other implements like python: 
* [tfrecord-lite](https://github.com/jongwook/tfrecord_lite/blob/b5207374801445f763967ae75fb1f798b71be758/tfrecord_lite.pyx#L64)

Or we can check the filesize and the current position from tell (also breaks when using  stdin or other streaming method)

**Comment by Leoyzen** [comment]
@jerry73204 It's quitely hard to distinguish these two situation. see this: read_exact and EOF.
I think it's normal behavior just quit the processing when there is breaking. I also saw this from other implements like python:

tfrecord-lite

Or we can check the filesize and the current position from tell (also breaks when using stdin or other streaming method)

**Comment by jerry73204** [comment]
The issue is fixed and release 0.8.0 is now available. Closing it.

**Pull request closed without merging**